writer/span_ring.gen.go

```diff
@@ -0,0 +1,257 @@
+	// slice high range.
+	lastHighStarted := 0
+	// The next index within the buffer that a datapoint should be added to.
+	nextDatapointIdx := 0
+	// Corresponds to nextDatapointIdx but is easier to work with without modulo
+	batched := 0
+	requestDoneCh := make(chan int64, w.MaxRequests)
+
+	// How many times around the ring buffer we have gone when putting
+	// datapoints onto the buffer
+	bufferedCircuits := int64(0)
+	// How many times around the ring buffer we have gone when starting
+	// requests
+	startedCircuits := int64(0)
+
+	targetHighStarted := func() int {
+		if nextDatapointIdx < lastHighStarted {
+			// Wrap around happened, just take what we have left until wrap
+			// around so that we can take a single slice of it since slice
+			// ranges can't wrap around.
+			return bufferSize
+		}
+
+		return nextDatapointIdx
+	}
+
+	tryToSendBufferChunk := func(newHigh int) bool {
+		if newHigh == lastHighStarted { // Nothing added
+			return false
+		}
+
+		if w.requestsActive >= int64(w.MaxRequests) {
+			w.requestsWaiting++
+			w.totalWaiting += int64(newHigh - lastHighStarted)
+			return false
+		}
+
+		count := int64(newHigh - lastHighStarted)
+		w.totalInFlight += count
+		w.requestsActive++
+
+		go func(low, high int) {
+			err := w.sendFunc(ctx, buffer[low:high])
+			if err != nil {
+				atomic.AddInt64(&w.totalFailedToSend, count)
+			} else {
+				atomic.AddInt64(&w.totalSent, count)
+			}
+
+			requestDoneCh <- count
+		}(lastHighStarted, newHigh)
+
+		lastHighStarted = newHigh
+		if lastHighStarted == bufferSize { // Wrap back to 0
+			lastHighStarted = 0
+			startedCircuits++
+		}
+
+		batched = 0
+		w.requestsWaiting = 0
+		w.totalWaiting = 0
+		return true
+	}
+
+	handleRequestDone := func(count int64) {
+		w.requestsActive--
+		w.totalInFlight -= count
+		if w.requestsWaiting > 0 {
+			tryToSendBufferChunk(targetHighStarted())
+		}
+	}
+
+	process := func(inst *trace.Span) {
+		w.totalReceived++
+
+		if w.preprocessFunc != nil && !w.preprocessFunc(inst) {
+			w.totalFilteredOut++
+			return
+		}
+
+		buffer[nextDatapointIdx] = inst
+
+		if lastHighStarted-1-int(w.totalInFlight) < nextDatapointIdx && bufferedCircuits > startedCircuits {
+			w.totalPotentiallyDropped++
+		}
+		batched++
+
+		nextDatapointIdx++
+		if nextDatapointIdx == bufferSize { // Wrap around the buffer
+			nextDatapointIdx = 0
+			bufferedCircuits++
+		}
+
+		if batched >= w.MaxBatchSize {
+			tryToSendBufferChunk(targetHighStarted())
+		}
+	}
+
+	waitForRequests := func() {
+		for w.requestsActive > 0 {
+			select {
+			case count := <-requestDoneCh:
+				handleRequestDone(count)
+			}
+		}
+	}
+
+	flush := func() {
+		newHigh := targetHighStarted()
+		// Could be less if wrapped around
+		if newHigh != lastHighStarted {
+			tryToSendBufferChunk(newHigh)
+		}
+	}
+
+	drainInput := func() {
+		defer waitForRequests()
+		defer flush()
+		for {
+			select {
+			case inst := <-w.inputChan:
+				process(inst)
+			default:
+				return
+			}
+		}
+	}
+
+	for {
```

Comment:
If I grok this right we favor incoming items (process), but will `flush` if nothing is in it. That checks the `newHigh` and sends along the buffer if needed. That sound right?